tests/ReCaptchaTest.php

```diff
@@ -0,0 +1,492 @@
+                ->setClient($client)
+                ->verifyOrThrow('test_token', '255.255.255.255');
+        } catch (FailedReCaptchaException $exception) {
+            $this->assertInstanceOf(ReCaptchaResponse::class, $exception->getResponse());
+
+            throw $exception;
+        }
+    }
+
+    public function testVerifiesReCaptchaDoesntThrowsException()
+    {
+        $client = $this->createMock(ClientInterface::class);
+
+        $array = [
+            'success' => true,
+            'error-codes' => [],
+            'hostname' => 'test.local.com',
+            'challenge_ts' => '1970-01-01T00:00:00Z',
+            'apk_package_name' => 'test_apk',
+            'score' => 1.0,
+            'action' => 'test_action',
+        ];
+
+        $stream = $this->factory->createStream(json_encode($array));
+
+        $stream->rewind();
+
+        $client->method('sendRequest')
+            ->with(new IsInstanceOf(RequestInterface::class))
+            ->willReturn($this->factory->createResponse()->withBody($stream)->withStatus(200, 'OK'));
+
+        $response = ReCaptcha::make('test_secret')
+            ->setClient($client)
+            ->verifyOrThrow('test_token', '255.255.255.255');
+
+        $this->assertTrue($response->valid());
+    }
+
+    public function testBuildsConstraints()
+    {
+        $recaptcha = ReCaptcha::make('test_secret');
+
+        $build = $recaptcha->hostname($hostname = 'test_hostname')
+            ->apkPackageName($apkPackageName = 'test_apk_package_name')
+            ->action($action = 'test_action')
+            ->threshold($threshold = 0.7)
+            ->challengeTs($challengeTs = rand(1, 120));
+
+        $this->assertInstanceOf(ReCaptcha::class, $build);
+
+        $array = [
+            'hostname' => $hostname,
+            'apk_package_name' => $apkPackageName,
+            'action' => $action,
+            'threshold' => $threshold,
+            'challenge_ts' => $challengeTs,
+        ];
+
+        $this->assertEquals($array, $recaptcha->getConstraints());
+        $this->assertEquals($array, $build->getConstraints());
+    }
+
+    public function testFlushesConstraints()
+    {
+        $recaptcha = ReCaptcha::make('test_secret')
+            ->hostname($hostname = 'test_hostname')
+            ->apkPackageName($apkPackageName = 'test_apk_package_name')
+            ->action($action = 'test_action')
+            ->threshold($threshold = 0.7)
+            ->challengeTs($challengeTs = rand(1, 120));
+
+        $this->assertEquals(ReCaptcha::CONSTRAINTS_ARRAY, $recaptcha->flushConstraints()->getConstraints());
+    }
+
+    public function testSanitizesAction()
+    {
+        $action = '/unsanitized@action/to-test_here?foo=bar&quz=qux';
+
+        $recaptcha = ReCaptcha::make('test_secret')
+            ->saneAction($action);
+
+        $this->assertEquals('/unsanitizedaction/to_test_here', $recaptcha->getConstraints()['action']);
+    }
+
+    public function testConstraintsErrors()
```

Comment:
```suggestion
    public function testConstraintsErrors(): void
```

could be used on all tests

Reply:
I don't think it's really necessary unless somebody would extend on it or there performance reasons behind it.

Reply:
I don't think it's really necessary unless somebody would extend on it or there performance reasons behind it.

Reply:
As you want :)